### PR TITLE
 AM-69: fixed post logout query param issue

### PR DESCRIPTION
**Pull request by ashraf706**
ParamUtils class moved to handler core as it is used by AbstractLogoutEndpoint.
Also added test for ParamUtils class which shows the URLs it can compare.

### CODE REVIEW

**Review by leleueri — CHANGES_REQUESTED** [review]
When the "allow wildcard" option is disable, the behaviour during signIn is not the same as the logout if my callback URL contains parameters .

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/logout/AbstractLogoutEndpoint.java [hunk 1]
```diff
@@ -235,9 +236,15 @@ private boolean isMatchingRedirectUri(String requestedRedirectUri, List<String>
...
-                .anyMatch(registeredUri -> requestedRedirectUri.equals(registeredUri));
+        if (isRedirectUriWildcardAllowed) {
+            return registeredRedirectUris
+                    .stream()
+                    .anyMatch(registeredUri -> ParamUtils.redirectMatches(requestedRedirectUri, registeredUri, false));
+        } else {
+            return registeredRedirectUris
+                    .stream()
+                    .anyMatch(registeredUri -> requestedRedirectUri.equals(registeredUri));
+        }
```
[the remark this answers is not in the document]
Comment (ashraf706) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/logout/AbstractLogoutEndpoint.java]: change the implementation to align with the behaviours of redirect_uris

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/root/resources/endpoint/logout/LogoutEndpointHandlerTest.java [hunk 2]
```diff
@@ -162,6 +162,33 @@ public void shouldInvokeLogoutEndpoint_targetUrl_client_restriction() throws Exc
                 HttpStatusCode.FOUND_302, "Found", null);
     }
 
+    @Test
```
Comment (leleueri) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/root/resources/endpoint/logout/LogoutEndpointHandlerTest.java:165]: Some TestCases are missing.

* isRedirectUriWildcardAllowed and valid target_url provided with parameter

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/root/resources/endpoint/ParamUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+package io.gravitee.am.gateway.handler.root.resources.endpoint;
+
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+public class ParamUtilsTest {
+
+    @Test
```
Comment (leleueri) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/root/resources/endpoint/ParamUtilsTest.java:25]: I think we can also add some tests with strictMatch required

**Review by leleueri — CHANGES_REQUESTED** [review]

**Comment by leleueri** [comment]
The CI is failing due to useless mock usage

File: gravitee-am-ui/src/app/domain/settings/openid/client-registration/settings/settings.component.html [hunk 4]
```diff
@@ -53,31 +53,31 @@
               [checked]="domain.oidc.clientRegistrationSettings.allowLocalhostRedirectUri" [disabled]="readonly">
               Allow localhost redirect uris.
             </mat-slide-toggle>
-            <mat-hint>Allow clients to be registered with redirect_uris using localhost as host.</mat-hint>
+            <mat-hint>Allow clients to be registered with redirect_uris/target_url using localhost as host.</mat-hint>
```
[the remark this answers is not in the document]
Comment (ashraf706) [reply gravitee-am-ui/src/app/domain/settings/openid/client-registration/settings/settings.component.html]: done

File: gravitee-am-ui/src/app/domain/settings/openid/client-registration/settings/settings.component.html [hunk 5]
```diff
@@ -53,31 +53,31 @@
...
           </div>
           <div class="domain-client-registration-settings-state" fxLayout="column">
             <mat-slide-toggle
               (change)="allowHttpSchemeRedirectUri($event)"
               [checked]="domain.oidc.clientRegistrationSettings.allowHttpSchemeRedirectUri" [disabled]="readonly">
               Allow unsecured (http) redirect uris.
             </mat-slide-toggle>
-            <mat-hint>Allow clients to be registered with redirect_uris using unsecured (http) scheme.</mat-hint>
+            <mat-hint>Allow clients to be registered with redirect_uris/target_url using unsecured (http) scheme.</mat-hint>
```
[the remark this answers is not in the document]
Comment (ashraf706) [reply gravitee-am-ui/src/app/domain/settings/openid/client-registration/settings/settings.component.html]: done

File: gravitee-am-ui/src/app/domain/settings/openid/client-registration/settings/settings.component.html [hunk 6]
```diff
@@ -53,31 +53,31 @@
...
           </div>
           <div class="domain-client-registration-settings-state" fxLayout="column">
             <mat-slide-toggle
               (change)="allowWildCardRedirectUri($event)"
               [checked]="domain.oidc.clientRegistrationSettings.allowWildCardRedirectUri" [disabled]="readonly">
               Allow wildcards redirect uris.
             </mat-slide-toggle>
-            <mat-hint>Allow clients to be registered with redirect_uris containing wildcards.</mat-hint>
+            <mat-hint>Allow clients to be registered with redirect_uris/target_url containing wildcards.</mat-hint>
```
[the remark this answers is not in the document]
Comment (ashraf706) [reply gravitee-am-ui/src/app/domain/settings/openid/client-registration/settings/settings.component.html]: done

File: gravitee-am-ui/src/app/domain/settings/openid/client-registration/settings/settings.component.html [hunk 7]
```diff
@@ -53,31 +53,31 @@
...
           </div>
           <div class="domain-client-registration-settings-state" fxLayout="column">
             <mat-slide-toggle
               (change)="enableRedirectUriStrictMatching($event)"
               [checked]="domain.oidc.redirectUriStrictMatching" [disabled]="readonly">
               Enable/Disable redirect uris strict matching.
             </mat-slide-toggle>
-            <mat-hint>Enable redirect_uris strict matching when calling OpenID Connect flow.</mat-hint>
+            <mat-hint>Enable redirect_uris/target_url strict matching when calling OpenID Connect flow.</mat-hint>
```
[the remark this answers is not in the document]
Comment (ashraf706) [reply gravitee-am-ui/src/app/domain/settings/openid/client-registration/settings/settings.component.html]: done

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/root/resources/endpoint/logout/LogoutEndpointHandlerTest.java (hunk 2 above)
Comment (ashraf706) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/root/resources/endpoint/logout/LogoutEndpointHandlerTest.java:165]: more tests added

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/root/resources/endpoint/ParamUtilsTest.java (hunk 3 above)
Comment (ashraf706) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/root/resources/endpoint/ParamUtilsTest.java:25]: test added

**Review by leleueri — APPROVED** [review]

**Comment by leleueri** [comment]
fixes gravitee-io/issues#8610

**Pull request merged**